guides/software-versioning.md

```diff
@@ -55,17 +55,37 @@ guidelines beyond it being the responsibility of the maintainer to keep an
 eye on how dependencies are evolving and set what version
 should be supported. However, two good rules of thumb are:
 
-* "The current version and the previous two" is often a reasonable choice of
-  dependency support. This is shorthand for "what the majority of the user
-  community uses", though, so if data on the user community's uptake is
-  available, that is preferred over this rule of thumb.
-* Only versions supported by the dependency's maintainers should be supported by
-  the dependent, unless the dependent software's maintainer is willing to take
-  complete responsibility for support of the old version of the dependency
+* "The current version and the previous two minor releases" is often a
+  reasonable choice of dependency support. This is shorthand for "what the
+  majority of the user community uses", though, so if data on the user
+  community's uptake is available, that is preferred over this rule of thumb.
+* Only versions supported by the dependency's maintainers should be supported
+  by the dependent, unless the dependent software's maintainer is willing to
+  take complete responsibility for support of the old version of the dependency
+
+Minimal pinning of dependencies should happen and should likely only be `>=`
+type of dependencies (e.g., `numpy>=1.12`) as opposed to pinning a dependency
+to a specific minor or patch release (i.e., don't use `numpy==1.12`). It is up
+to the package maintainer to regularly check for new releases of dependencies
+and to confirm they are working (this could be by automated weekly cron jobs,
+for example). If a dependency creates a release that breaks the maintainer's
+package, then the maintainer should create a point-release and pin an upper
```

Comment:
I'm not so sure about this "should" here.  I think in general it's ok to accept that sometimes dependencies change faster than the dependent, and it's not on us if a dependency is lax about deprecation - that's *their* fault for not following good practice, so the burden doesn't need to be on us.  That is, there's not necessarily a need to release a point release *just* to tell people that numpy 1.13 is incompatible.  Relatedly, it may be dangerous in general to do the <= sort of pinning because what if 1.13.1 fixes it?  Do we need yet *another* point release?  

So I'd say this should be softened to something like "at the next convenient release", and also maybe encourage "exclusion" releases (e.g. !=1.13 instead of <=1.13)?  Of course this is situation in that sometimes it's known that it's a *permanent* break (i.e., a feature removed) instead of a bug, in which case <= 1.13 is warranted... So maybe it should just be something like ``to pin "numpy<1.13" or "numpy!=1.13.0"``?

Reply:
At Astropy, we briefly did a `numpy>=1.11,numpy!=1.13`. That syntax works for `pip` (not sure about `conda`).

Reply:
Ah... hmmmm, but then if people are installing via PyPi they could very well install a broken package. I think the onus is on us to make sure that pip install's will work, no?

One concrete example is glueviz and cubeviz. When glueviz 0.15 is released then anyone pip installing cubeviz will get a broken package. That is going to make a lot of frustrated users, no? Then, if the user has to go through and figure out why cubeviz is not working and *maybe* will get to the point that they need to downgrade glueviz to 0.14 -- seems like a lot of potential time and frustration on the users part, no?  

It seems like a simple point release (e.g., cubeviz) that does some sort of upper pin (e.g., on glueviz) would alleviate a lot of the potential frustration for the users doing installations. Or, as you suggest a `!=` pin if it seems like something will be fixed.

Maybe it comes down to whether the dependency is past a major release (ie > 1.0) or not.  Something like numpy will be adhering more to major, minor and point releases vs a very fast moving package like glueviz which still has a version < 1.0.

Reply:
I tend to think viz + glue case as the exception, not the norm?

Reply:
In general I agree with @eteq, but there will be cases where a dependency will have a major release or introduce breaking changes without a deprecation period. In these cases I think it's perfectly reasonable to issue a point release with a maximum version restriction, with the understanding that we will need to fix the underlying compatibility issue eventually.

This is an extreme case, but consider the migration from Py27 to Py3. There were many packages that had a good reason to continue maintenance of exclusive Py27 code bases for quite a while, even though obviously they had to update eventually. Or the migration from PyQt4 to PyQt5. In these cases, it's okay to explicitly restrict the maximum compatible version for a defined period of time since transition efforts can be quite burdensome and it's necessary to support functional software in the meantime.

However, I agree that we need to be as flexible as possible with packages like `numpy` and `scipy` and avoid issuing maximum restrictions as much as possible. And I agree with @eteq that in cases where  e.g. `numpy` publishes a broken point release, we are not under any obligation to release a bugfix of our packages to avoid that release. We should just advise users of the problem and wait until `numpy` gets fixed upsrtream.

Reply:
@drdavella I'm not sure I completely understand what you think we ought to do in the case of a broken point release of a major upstream dependency: 

> in cases where e.g. `numpy` publishes a broken point release, we are not under any obligation to release a bugfix of our packages to avoid that release.

It seems like this is a case where we need to do something once we know there is a problem. The quick "bug fix" for the problem is to indicate that our software doesn't support that version of `numpy` -- preferably in an automated way so pip and conda will help the user solve the dependency problem. Then if they fix it upstream and we don't need to revise our software, does that sort of "automated version exclusion" just become permanent, or can they somehow withdraw that release so we don't have to worry about excluding it on our end?

Reply:
AFAIK no one upstream ever withdraws a release. One can also opt to just document it in a "known issues" section of the user facing doc, as Astropy has done for some of the compatibility issues.

Reply:
Documenting as a known issue seems fine, but better if we can automate the build process (like conda) so that it will know that it has to downgrade or upgrade the dependency to be able to build a working version.

Reply:
@hcferguson the case I have in mind is where e.g. `numpy` issues a point release with a bug, then they quickly discover it's a bug, and then they quickly release a bugfix. 

In this case, it's not even guaranteed that any of our users would notice the bug. They would have to update `numpy` for some reason and then run our software. If they **did** notice the bug, and they opened an issue in one of our repos, then we can simply respond to the issue with a comment about downgrading `numpy` until the next bugfix release is available. I don't think this kind of scenario should drive a point release on our end.

In this case it might still be useful to open a PR on our end in order avoid the buggy dependency version in the future, but I still don't think that PR should drive a new release. It can just be included in the next scheduled release.

You can imagine variants of this scenario where the upstream dependency doesn't get fixed quickly and it ends up affecting a large number of users. Even in that case, we might consider   issuing a new release ourselves, but the effect is going to be the same: we're going to ask the end users to install a new version of one piece of software or another (either ours or the dependency) until the bug is resolved.

Maybe the takeaway from this discussion is that it's pretty hard to write general guidance for these kinds of scenarios. We need to be flexible and able to make decisions on a case-by-case basis. But I would still argue that in the general case we are not obligated to release simply to avoid a buggy upstream version.

Reply:
Also, I don't mean to pick on `numpy`, but they do release pretty often and almost all of our software depends on it. So when these kinds of things happen, chances are good that we're talking about `numpy`.

Reply:
Or `pytest`... :wink:

Reply:
or conda... oh wait, that is not a dependency, I am so used to thinking conda might be related to "long discussions" we have sometimes  ;)

Reply:
@drdavella Thanks for the clarification. Seems sensible. My only caution is that one often doesn't know how quick the bugfix is going to be. But I agree we aren't obligated to make a new release the instant we discover the problem.

Reply:
@eteq Given the discussion here, is there still something you would like changed, or are you good with the text (and can we merge)?

Reply:
@brechmos-stsci It looked to me like the consensus is that the statement is too strong here, in particular, it looked like there was agreement that the "upper bound" is not a good idea unless absolutely necessary.

So what about:
```suggestion
package, then the maintainer should create an issue in the dependency's issue tracker reporting a problem.  If not promptly fixed, an exclusion of the broken dependency version (i.e. one that uses the exclusion mechanics of [PEP 440](https://www.python.org/dev/peps/pep-0440)) should be added in the next release of the maintained package, or if impractical, a work-around. If fixed quickly enough, an exclusion may not be necessary, although a "known issues" note may be useful if such a section exists in the documentation or changelog.
```
(That's really a replacement for the rest of the paragraph... but there's not an easy way to do multi-line suggestions)

```
bound on the dependency and create an issue in their package to create a
work-around for the dependency's new release. For example, if numpy 1.14 is
released and results in a bug in our package then the point-release should be
created to pin "numpy<=1.13" and an issue be created in the package to fix the
error. It is not *expected* that this should happen as dependencies should use
deprecation so there is a period of time over which changes can be made.
```

Reply:
Oh and I guess the "for example" could be updated to reflect the above suggestion?